data-pipeline-flink/device-profile-updater/src/main/scala/org/sunbird/dp/deviceprofile/functions/DeviceProfileUpdaterFunction.scala

```diff
@@ -0,0 +1,198 @@
+
+    preparedStatement.executeUpdate
+    preparedStatement.close()
+    // Update first_access column if the first_access is null
+    val updateFirstAccessQuery = String.format("UPDATE %s SET first_access = '%s' WHERE device_id = '%s' AND first_access IS NULL", config.postgresTable, new Timestamp(firstAccess).toString, deviceId)
+    postgresConnect.execute(updateFirstAccessQuery)
+
+    if (null != deviceData.get(config.userDeclaredState)) {
+      // Updating user_declared_on column
+      val updateUserDeclaredOnQuery = String.format("UPDATE %s SET user_declared_on = '%s' WHERE device_id = '%s' AND user_declared_on IS NULL", config.postgresTable, new Timestamp(lastUpdatedDate).toString, deviceId)
+      postgresConnect.execute(updateUserDeclaredOnQuery)
+    }
+    logger.info(s"Device ID: $deviceId Updated the postgres device profile table")
+  }
+
+  private def addDeviceDataToCache(deviceId: String, deviceProfile: DeviceProfile): util.Map[String, String] = {
+    val deviceMap = deviceProfile.toMap(config)
+    deviceMap.values.removeAll(Collections.singleton(""))
+    deviceMap.values.removeAll(Collections.singleton("{}"))
+    if (deviceMap.get(config.userDeclaredState) == null) deviceMap.remove(config.userDeclaredOn)
+    if (dataCache.isExists(deviceId)) {
+      val redisData = dataCache.hgetAllWithRetry(deviceId)
+      val firstAccess = redisData.get(config.firstAccess)
+      val userDeclaredOn = redisData.get(config.userDeclaredOn)
+      if (firstAccess != null && "0" != firstAccess) deviceMap.remove(config.firstAccess)
+      if (userDeclaredOn != null && deviceMap.get(config.userDeclaredOn) != null) deviceMap.remove(config.userDeclaredOn)
+      val updatedDeviceMap = updatedMissingFields(deviceMap, redisData)
+      dataCache.hmSet(deviceId, updatedDeviceMap.asInstanceOf[util.Map[String, String]])
+      logger.info(s"Device ID: $deviceId is present in the redis and required fields into redis")
+      updatedDeviceMap
+    }
+    else {
+      dataCache.hmSet(deviceId, deviceMap.asInstanceOf[util.Map[String, String]])
+      logger.info(s"Device ID: $deviceId does not exists in the redis and all fields into redis")
+      deviceMap
+    }
+  }
+
+  @throws[SQLException]
+  private def setPrepareStatement(preparedStatement: PreparedStatement, index: Int, deviceData: util.Map[String, AnyRef]): Unit = {
+    val gson = new Gson()
+    var count = index
+    deviceData.values().forEach(value => {
+      count += 1
+      val jsonObject = new PGobject
+      try {
+        jsonObject.setType("json")
+        jsonObject.setValue(gson.fromJson(value.toString, classOf[JsonObject]).toString)
+        preparedStatement.setObject(count, jsonObject)
+      } catch {
+        case ex @ (_: JsonSyntaxException | _: ClassCastException) =>
+          preparedStatement.setString(count, value.toString)
+      }
+    })
```

Comment:
There are some fixes done here to handle string types. Is it considered?

Reply:
Yes, Santhosh It's considered, You mean to handle the special char in the strings?
 For example : 
`"user_declared_district": "Bengaluru's",
  "user_declared_state": "Karnataka,'s"`
It's able to insert the data and wrote testcase for special char as well.